mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,271 @@ def load_model(model_uri):
+        """
+        value = (eval_dataset_name, model._mlflow_uuid)
+        prediction_result_id = id(predict_result)
+        self.pred_result_id_to_dataset_name_and_model_id[prediction_result_id] = value
+
+        def clean_id(id_):
+            _get_autolog_training_status().pred_result_id_to_dataset_name_and_model_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+        return predict_result
+
+    def register_metric_api_call(self, metric_fn, call_pos_args, call_kwargs):
+        """
+        Given a metric api call (include the called metric function, and call arguments)
+        Register the call information (arguments dict) into the `metric_api_call_arg_dict_list_map`
+        and return a tuple of (metric key, run_id):
+        - metric key format is: {metric_name}_on_{eval_dataset_name}
+          metric_name is generated by metric function name, if there're multiple calls with
+          different call arguments, then appending index to the name.
+        - run_id is the id of mlflow run which we need log metric into.
+        """
+        call_arg_list = list(call_pos_args) + list(call_kwargs.values())
+
+        dataset_id_list = self.pred_result_id_to_dataset_name_and_model_id.keys()
+        print(f'DGB: get_eval_dataset_name: # 1')
+
+        dataset_name = None
+        model_id = None
+
+        # Note: some metric API the arguments is not like `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        for arg in call_arg_list:
+            if arg is not None and not np.isscalar(arg) and id(arg) in dataset_id_list:
+                print(f'DGB: get_eval_dataset_name: # 3')
+                dataset_name, model_id = self.pred_result_id_to_dataset_name_and_model_id[id(arg)]
+                break
+
+        if dataset_name is None or model_id is None:
+            return None
+
+        run_id = self.model_id_to_run_id_map.get(model_id, None)
+        if run_id is None:
+            return None
+
+        call_arg_dict = _extract_metric_api_call_arg_dict(metric_fn, call_pos_args, call_kwargs)
+        metric_name = metric_fn.__name__
+
+        registered_call_arg_dict_list = self.metric_api_call_arg_dict_list_map[metric_name]
+
+        index = len(registered_call_arg_dict_list)
+        for i, call_arg_dict_i in enumerate(registered_call_arg_dict_list):
+            if call_arg_dict == call_arg_dict_i:
+                index = i
+                break
```

Comment:
`call_arg_dict` doesn't contain dataset arguments, right? Because of that, what happens if I pass different boolean or string arguments to a metric evaluation function? For example,

```
sklearn.metrics.top_k_accuracy_score(y_true, y_score, k=1, normalize=True)
sklearn.metrics.top_k_accuracy_score(y_true, y_score, k=1, normalize=False)
```

Will this produce two separate metrics or overwrite one metric? Because it's probably unlikely that users will call the exact same metric evaluation function twice, is it okay if we always increment the index and log a new metric each time?

Reply:
see https://github.com/mlflow/mlflow/pull/4491#discussion_r670975352
it will be logged as two different metric key